cmd/snap/cmd_debug_state.go

```diff
@@ -293,6 +301,165 @@ func (c *cmdDebugState) showIsSeeded(st *state.State) error {
+	Undesired        bool                   `json:"undesired,omitempty" yaml:"undesired"`
+	StaticPlugAttrs  map[string]interface{} `json:"plug-static,omitempty" yaml:"plug-static,omitempty"`
+	DynamicPlugAttrs map[string]interface{} `json:"plug-dynamic,omitempty" yaml:"plug-dynamic,omitempty"`
+	StaticSlotAttrs  map[string]interface{} `json:"slot-static,omitempty" yaml:"slot-static,omitempty"`
+	DynamicSlotAttrs map[string]interface{} `json:"slot-dynamic,omitempty" yaml:"slot-dynamic,omitempty"`
+	// TODO: hotplug
+}
+
+type connectionInfo struct {
+	PlugSnap string
+	PlugName string
+	SlotSnap string
+	SlotName string
+
+	connState
+}
+
+type byPlug []*connectionInfo
+
+func (c byPlug) Len() int      { return len(c) }
+func (c byPlug) Swap(i, j int) { c[i], c[j] = c[j], c[i] }
+func (c byPlug) Less(i, j int) bool {
+	a, b := c[i], c[j]
+	return a.PlugSnap < b.PlugSnap || (a.PlugSnap == b.PlugSnap && a.PlugName < b.PlugName)
+}
+
+func (c *cmdDebugState) showConnectionDetails(st *state.State, connArg string) error {
+	st.Lock()
+	defer st.Unlock()
+
+	p := strings.FieldsFunc(connArg, func(r rune) bool {
+		return r == ' ' || r == ','
+	})
+
+	var plugMatch, slotMatch SnapAndName
+	if err := plugMatch.UnmarshalFlag(p[0]); err != nil {
+		return err
+	}
+
+	if len(p) > 1 {
+		if err := slotMatch.UnmarshalFlag(p[1]); err != nil {
+			return err
+		}
+	}
+
+	var conns map[string]*connState
+	if err := st.Get("conns", &conns); err != nil && !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+
+	// sort by connection ID
+	connIDs := make([]string, 0, len(conns))
+	for connID := range conns {
+		connIDs = append(connIDs, connID)
+	}
+	sort.Strings(connIDs)
+
+	for _, connID := range connIDs {
+		connRef, err := interfaces.ParseConnRef(connID)
+		if err != nil {
+			return err
+		}
+
+		refMatch := func(x SnapAndName, y interface{ String() string }) bool {
+			parts := strings.Split(y.String(), ":")
+			return len(parts) == 2 && x.Snap == parts[0] && x.Name == parts[1]
+		}
+		plug, slot := connRef.PlugRef, connRef.SlotRef
+
+		if slotMatch.Name != "" && slotMatch.Snap != "" && plugMatch.Snap != "" && plugMatch.Name != "" {
+			// command invoked with 'snap:plug,snap:slot'
+			if !refMatch(plugMatch, plug) || !refMatch(slotMatch, slot) {
+				// should match the connection exactly
+				continue
+			}
+		} else if plugMatch.Snap != "" && plugMatch.Name != "" && slotMatch.Snap == "" && slotMatch.Name == "" {
+			// command invoked with 'snap:plug-or-slot'
+			if !refMatch(plugMatch, plug) && !refMatch(plugMatch, slot) {
+				// should match either the connection's slot or plug
+				continue
+			}
+		} else if plugMatch.Snap != "" && plugMatch.Name == "" && slotMatch.Snap == "" && slotMatch.Name == "" {
+			// command invoked with 'snap' only
+			if plugMatch.Snap != slot.Snap && plugMatch.Snap != plug.Snap {
+				// should match one of the snap names
+				continue
+			}
+		} else {
```

Comment:
Don't we miss a case where only the slot is given? I'm probably confused, because I see that in the test `TestDebugConnectionDetails` we are testing the case where the parameter is `",gtk-common-themes:gtk-3-themes"` which I guess should result in an empty plugMatch and in an valid slotMatch; but according to this series of `if`s, that case should land into this `else` branch and cause an error -- yet the test is passing.

Please enlighten me :-)

Reply:
Yes, that makes sense, but then isn't the `TestDebugConnectionDetails()` test testing the `snap:slot` case, when it runs `snap debug state --connection=,gtk-common-themes:gtk-3-themes`?

Reply:
It's testing the `snap:name` case, since there is no way to print connections specifically of slots matching some snap. `snap:name` prints any connection matching that name, regardless of whether it is a plug or a slot. Based on my understanding of the previous state of the PR, that was the original intent. Did I understand your question correctly?

Reply:
The thing that confuses me in that test is the comma in `--connection=,gtk-common-themes:gtk-3-themes`: looking at the code at the beginning of `showConnectionDetails()`, the `strings.FieldsFunc()` function is used to split the argument string either at whitespace or at commas, and the resulting parts are interpreted as plug match (`p[0]`) and slot match (`p[1]`, if present). So, from here it would seem that by passing `,gtk-common-themes:gtk-3-themes` the test intent was to set `p[0]` to an empty string and `p[1]` to `gtk-common-themes:gtk-3-themes`. So, we'd get an empty plug match, and a non empty slot match.

My guess is that the whole point of allowing the comma as a separator was to allow the user to specify a match on the slot. I can't say whether this is worthwhile, but (unless my understanding is wrong) I think we should either support this case, or drop it completely.

Reply:
> My guess is that the whole point of allowing the comma as a separator was to allow the user to specify a match on the slot.

I don't think that was the intent but I may be interpreting it incorrectly. I think the comma is just there to test that it doesn't break anything but it doesn't mean "match this as a slot". The command documentation is `"Show details of the matching connections (snap or snap:plug,snap:slot or snap:plug-or-slot")` which doesn't include `,snap:slot`. I think if that was the intent, it would've been documented. The `strings.FieldFunc` splits strings on separators but doesn't include empty strings in the output, so commas at the end or not separated by other runes are skipped over. So it's impossible for `p[0]` to be empty and `p[1]` to be non-empty. Even if it could be empty, `if err := plugMatch.UnmarshalFlag(p[0]); err != nil {` would always fail, because it the unmarshalling fails if both the snap and name are empty. If you look below that, the slot parsing has this guard `if len(p) > 1 {` so it would seem that the plug can never be empty but the slot can. So it seems to me that the current code never intended to support that. The test was just adding it to make sure it didn't break anything.
To be clear, I'm not opposed to supporting it now, it just requires some extra code :slightly_smiling_face:

Reply:
I see, I didn't know that empty fields were skipped. I'm not sure I like the asymmetry in the UX (that you can filter for a plug, but not for a slot) and the test with the initial comma looks very suspicious, but given that this is a debug command I think I shouldn't be too picky :-)

Reply:
Just one tiny note to make sure I didn't say anything misleading: it's not that you can filter for a plug and not a slot. You can either filter for both at the same time or, if there's just one name, you can filter for a connection name (regardless of whether it's a plug or a slot).

Reply:
Ah, right, I just got confused by reusing the plugMatch variable, but the code itself is clear and well commented, so all is fine :-)